tmt/steps/provision/__init__.py

```diff
@@ -2507,77 +2564,96 @@ def push(
         superuser: bool = False,
     ) -> None:
         """
-        Push files to the guest
-
-        By default the whole plan workdir is synced to the same location
-        on the guest. Use the 'source' and 'destination' to sync custom
-        location and the 'options' parameter to modify default options
-        which are '-Rrz --links --safe-links --delete'.
-
-        Set 'superuser' if rsync command has to run as root or passwordless
-        sudo on the Guest (e.g. pushing to r/o destination)
+        Push files or directories to the guest using rsync.
         """
-
-        # Abort if guest is unavailable
         if self.primary_address is None and not self.is_dry_run:
-            raise tmt.utils.GeneralError('The guest is not available.')
-
-        # Prepare options and the push command
-        options = options or DEFAULT_RSYNC_PUSH_OPTIONS
-        if destination is None:
-            destination = Path("/")
-        if source is None:
-            # FIXME: cast() - https://github.com/teemtee/tmt/issues/1372
-            parent = cast(Provision, self.parent)
+            raise GeneralError('The guest is not available.')
 
-            assert parent.plan.workdir is not None
-
-            source = parent.plan.workdir
-            self.debug(f"Push workdir to guest '{self.primary_address}'.")
-        else:
-            self.debug(f"Copy '{source}' to '{destination}' on the guest.")
+        current_rsync_options = options or DEFAULT_RSYNC_PUSH_OPTIONS
 
-        def rsync() -> None:
-            """
-            Run the rsync command
-            """
+        try:
+            source_str = self._prepare_rsync_source_path(source)
+            dest_str = self._prepare_rsync_destination_path(destination)
+        except ValueError as path_prep_error:
+            raise GeneralError(f"Failed to prepare rsync paths: {path_prep_error}")
 
-            # In closure, mypy has hard times to reason about the state of used variables.
-            assert options
-            assert source
-            assert destination
+        self.debug(
+            f"Resolved rsync paths: source='{source_str}',"
+            f" destination='{dest_str}' for guest '{self.primary_address}'"
+        )
 
-            cmd = ['rsync']
+        def _execute_rsync_operation() -> None:
+            cmd_base = ['rsync']
             if superuser and self.user != 'root':
-                cmd += ['--rsync-path', 'sudo rsync']
+                cmd_base.extend(['--rsync-path', 'sudo rsync'])
+
+            ssh_command_str = self._ssh_command.to_element()
+            remote_target = f"{self._ssh_guest}:{dest_str}"
+
+            full_cmd_list = [
+                *cmd_base,
+                *current_rsync_options,
+                "-e",
+                ssh_command_str,
+                source_str,
+                remote_target,
+            ]
 
+            self.debug(f"Executing rsync: {' '.join(repr(arg) for arg in full_cmd_list)}", level=2)
```

Comment:
Logging is done by `Command.run()` under the hood of `self._run_guest_command()`, and `full_cmd_list` does not need to exist and can be expanded in `Command(...)` below...